### ISSUE TITLE
Version 4.3 coming soon

### ISSUE THREAD

**Issue by paula-w** [YetiForceCompany/YetiForceCRM#5526]
With version 4.3 just around the corner, there are a few things we would like to say.

First of all, we would like to thank **you** - our amazing community - for helping us make YetiForce better. We appreciate all your efforts, and we are grateful for every single pull request, bug report, and suggestion you sent us. 

In this issue, you will find the link to the RC version. This version includes many fixes and improvements (changelog in the first comment), however, we are aware that we might have missed a bug or two, and we would like to ask you to help us find these bugs.

If you want to report a bug that you found, please provide us with a very detailed step-by-step description how to duplicate it, and send your logs to github@yetiforce.com (the subject of the email must be the GitHub issue number)  

You can download the update package from: https://github.com/YetiForceCompany/UpdatePackages/tree/developer/YetiForce%20CRM%204.x.x/4.2.0_to_4.3.0/zip (package updated on 07.12.2017) 

### IMPORTANT: 
- The update package introduces a new password encryption mechanism that requires setting a new password. We recommend enabling the reset password mechanism when performing the update and configure SMTP. 
There is an option for the system to reset system administrators’ passwords and send them via email. 
If the passwords are not mailed we recommend setting a new password for all users after the update.

- All widgets in record summary should be configured once again after the update ( Software configuration > Standard modules > Modules – Widgets); the configuration has been rebuilt, which will cause the widgets to stop working. Please open the widget and save it again.

**Comment by paula-w** [comment]
**ADDED**

- Added verification of permissions
- Added a generator for user and calendar colors
- Added picklist css generator
- Added verification of dangerous usernames
- Added ability to change username during installation
- Added enhancements to the system installer
- Added username verification during installation
- Added new charts to widgets
- Added filtering by owner to widgets with charts
- Added advanced verification of phone numbers
- Added a new field - Country
- Added new graphs to widgets
- Added testing of a record list
- Added verification of data types while saving
- Added summation to charts
- Added testing of the record list
- Added tests for API integration
- Added a new functionality to reset a user’s password
- Added a new functionality to change a user’s password
- Added mass password reset
- Added new tests
- Added a new functionality to remember previous user’s passwords
- Added brute force to reset password
- Added user password lifetime
- Added configuration of user account lock time
- Added functionality to force user password change 
- Added a new filter for users 
- Added functionality for archiving records
- Added coloring for record states
- Added style enhancements
- Added confirmations to actions
- Added validation of entered data to inventory table


**IMPROVED**

- Changed the license to YetiForce Public License v3
- Improved error handling
- Improved TextParser
- Improved editor of Roundcube Configuration
- Reconstruction of group functions
- Improved the panel for the search configuration
- Improved mail scanner
- Updated user colors
- Updated composer.json
- Updated .gitignore
- Updated phpunit.xml
- Updated .travis.yml
- Updated README.md
- Improved security
- Improved the quick create form for calendar
- Improved sql queries
- Improved automatic loading of values
- Improved create the menu
- Moved colors to the new settings module
- Rebuilt handling errors and clearing input
- Rebuilt test structure
- Reconstructed input data security
- Improved view of event information
- Improved phone number storage
- Improved translation mechanism
- Improved permissions to delete records
- Improved cache of data in installation
- Improved configuration saving in ApiAddress module
- Improved email binding
- Improved code quality
- Improved security of input data
- Improved security exceptions 
- Improved functions of dates
- Improved data cleaning
- Improved permissions for Custom View
- Improved the list of mass actions
- Improved the login functionality
- Improved record deletion


**FIXED**

- Fixed loading links
- Fixed a bug that appeared after opening an event in list view
- Fixed a bug that appeared when creating a filter
- Fixed calendar colors in configuration
- Fixed attaching emails in contacts
- Fixed an advanced filter condition
- Fixed filter conditions
- Fixed a deletion of email scanner accounts
- Fixed errors in permissions
- Fixed create user records
- Fixed displaying of user image
- Fixed import
- Fixed folder conditions on folder and email fields
- Fixed tests
- Fixed group and role creation
- Fixed deletion of groups
- Fixed workflows navigation 
- Fixed a function call
- Fixed deleting users
- Fixed add relation in quick create
- Fixed field validation
- Fixed global search
- Fixed quick create error
- Fixed the FAQ module
- Fixed a bug that didn’t allow to add comments
- Fixed validation of url
- Fixed adding events
- Fixed the display of working time in Competition
- Fixed Campaigns reports
- Added minor fixes to phone field
- Fixed compatibility functions
- Changed the password encryption algorithm
- Fixed an error in OSSPassword
- Fixed system installer
- Fixed bugs in Reports
- Fixed mass edit in Users
- Fixed bugs in Workflows
- Fixed a bug that didn’t allow to add or remove new widgets in modules
- Fixed a bug that didn’t allow to use advanced search when in the module
- Fixed a bug with permissions in profiles
- Fixed fields updating in the Geolocator
- Fixed Email Scanner
- Fixed related documents so that they become clickable
- Fixed a bug that appeared during generating a PDF Template
- Fixed mass SMS
- Fixed a fatal error in advanced block fields
- Fixed Contact relation with Sold Products and Services

**REMOVED**

- Removed PearDatabase
- Removed GetParentGroups.php file
- Removed moduleColor css class
- Removed mt940 import
- Removed EditFieldByModal
- Removed redundant code
- Removed DataAccess module
- Removed an old password reset and password change functionality

**Comment by PercyP** [comment]
Hi @paula-w 

First of all CONGRATULATIONS - its obvious so much work has gone in to this (as always).

Just downloaded and installed upgrade without any problems, thought mine would cause many issues having customised it so much. Will now start testing it and feed back any issues/bugs.

Looks amazing - loving all the new 'unexpected' features. Well impressed 👍 Cannot wait to explore more.

**Comment by kpaulaha** [comment]
Installation went ok, the custom modules made in 4.2 not working any more. Must inspect tomorrow. Otherwise seems fine. Great.

**Comment by PercyP** [comment]
I too am finding a couple of small issues. I use the Competition module for another purpose and this no longer works - I get an error of: 

Incorrect request ERR_NOT_ALLOWED_VALUE

When I click on the menu link for the module. 

I too will investigate properly tomorrow and check other parts of the system too

**Comment by zebrahosting** [comment]
Some feedback on the RC version. 
Install went well if PHP timeout is set long enough.
Checking the version number still gives 4.2.514 in the footer after the update.
Checking the modules, the Roundcube lib needs updating, pressed the button and it became red (from yellow) needed to download. Pressed the download button but nothing happened.
Did a manual download here from github and tried to install as file but failed. (Module does not contain default language (en_us).)
Custom module not working anymore
Our filter issues still there.

**Comment by kpaulaha** [comment]
Well the custom modules works ok, the menu not, remowing "&mid=359&parent=0" from the menu address "index.php?module=OSSEmployees&view=List&mid=359&parent=0" opens it.

**Comment by kpaulaha** [comment]
Adding new Workflow item not working, New Workflow -button has no effect.
![workflow system settings](https://user-images.githubusercontent.com/10330264/32980820-56180934-cc76-11e7-885e-e2ad05dc260c.png)

**Comment by PercyP** [comment]
@kpaulaha 

how were you able to test that your custom module worked ok? Is it just the menu links that don't work in your case? I am setting some time aside later (or first thing tomorrow) to go through my update and feed back.

**Comment by zebrahosting** [comment]
Same here. Removed and added the menu but keep getting:
Incorrect request
Incorrect value detected, please contact your administrator

**Comment by kpaulaha** [comment]
@PercyP 
I made a vidget on the home page viewing records of the custom module, and it did show. Opened from that a record and to the list view. Have 4 custom modules and all works fine, only the url from menu does't work, not even if you make a new menu item.

**Comment by PercyP** [comment]
Hi,
it appears the problem is only with main links in the side menu. If you move a link to a sub menu it works. For example I had the Email link as a main menu item but then I moved it under Contacts and it opens fine. Other Links that have no sub-menus i.e. they are opened direct from the menu produce the same error - see attached screenshot
![errormessage](https://user-images.githubusercontent.com/10598886/32989857-6cd8dc3e-cd16-11e7-8943-91fe93c0329a.png)

**Comment by PercyP** [comment]
I just tested on developer version. 

I created a new menu item called Test Account and added it to the bottom of the menu. Then I tried to open it and get the following error message:
Incorrect request
Incorrect value detected, please contact your administrator

#0 vendor/yetiforce/Request.php(136): App\Purifier::purifyByType('0', 1)
#1 include/runtime/Controller.php(299): App\Request->getByType('parent', 1)
#2 modules/Vtiger/views/Basic.php(30): Vtiger_View_Controller->preProcess(Object(App\Request), false)
#3 modules/Vtiger/views/Index.php(34): Vtiger_Basic_View->preProcess(Object(App\Request), false)
#4 modules/Vtiger/views/List.php(64): Vtiger_Index_View->preProcess(Object(App\Request), false)
#5 include/main/WebUI.php(258): Vtiger_List_View->preProcess(Object(App\Request))
#6 include/main/WebUI.php(189): Vtiger_WebUI->triggerPreProcess(Object(Vtiger_List_View), Object(App\Request))
#7 index.php(25): Vtiger_WebUI->process(Object(App\Request))
#8 public_html/index.php(11): require('/home/gitdevelo...')
#9 {main}
#0 [trace] Entering getColumnFields(Users) method ...
#1 [trace] Exiting getColumnFields method ...
#2 [info] Opening DB connection: mysql:host=localhost;dbname=gitdevelop_crm;port=3306 || yii\db\Connection::open
#3 [info] INSERT INTO `u_yf_browsinghistory` (`userid`, `date`, `title`, `url`) VALUES (5, '2017-11-19 11:49:18', 'Accounts List', 'index.php?module=Accounts&view=List&mid=152&parent=0') || yii\db\Command::execute
#4 [error] purifyByType: 0 || BadRequest
#5 [error] ERR_NOT_ALLOWED_VALUE||0 => vendor/yetiforce/Purifier.php:326

See attached screenshot of the same:
![gitdeveloper](https://user-images.githubusercontent.com/10598886/32989919-6c26752a-cd17-11e7-876b-80c12abd6e64.png)

**Comment by paula-w** [comment]
Thank you all a lot for all the errors you found, and for all the effort you put into finding them during the weekend.
Could we ask you to open separate issues for each of these errors, with a description how to replicate them and error logs? You can add in the description that these errors popped up after the update and you can reference this issue to avoid duplicates. It's going to make it much easier for us to track and solve the issues. 

@zebrahosting Roundcube update should work now, thanks!

**Comment by migoi** [comment]
Please don't get me wrong, but this are too much (fundamental) changes in a minor version update.

This time are a lot of code breaking changes inside. If you have custom modules, there is no way to follow all those steps. Those step to replace code "old style" to "new style" should be documented, so anyone can follow without reading thousands of commits. At least when inside a minor version number update.

I know and understand most of them are because of security reasons, but that deserves a major version step and so a "real" update plan to get up with code changes.

Fundamental functions, that no longer exists should get a "deprecated" time frame. So a update from 4.x to 4.y would not harm an existing installation.

Sometimes less is more.

**Comment by paula-w** [comment]
This is @bpabiszczak's answer:

If the updates are too frequent or too hard for some people then they can update their system less frequently [eg. every 2 – 3 versions] or order the update from someone who can do it for them. You have to remember that each change in the system affects us more than anybody else because we update several dozen systems. However, we think these changes are good and necessary - this is why we make them.

**Comment by migoi** [comment]
no, not less, better more frequent. Right now even fatal error will only be "official" fixed after 3-4 month. Of course you can patch your system for yourself and have it fixed immediately. But that's quite hard to follow if you are not the reporter of the failure/fix.

And yes: it will if affect you even more. So with more and more success to this product, this problem will become greater and greater.

And even more yes: these changes are necessary - absolutely. The code base is grown over years, From the code itself you can see ancient stuff, bad stuff and many lines you would never ever code like this. There is a clean-up needed - no question. The problem is not the improvement itself, it's the way how you can get it.
Maybe on possible solution is to rethink the release cycle itself, to make more intensive use of the git branches possibility, so that such a "fatal fix" can be (auto-) released within a day to the stable branch and you continue to work on the "next" branch. Sa the server's will updates themselves or can be updated by a simple click to the latest stable. Instead a big update a lot of small updates.

All I want to say is, that its quite hard to follow the development. And I also know how hard it is to communicate the current development actions to other (remote) developers. That's one of the core problems each open source project has. 

E.g. Can we have "developer chat hour" once in a month?
Or Can we have a reference module, so others can see what is the "state of the art" access to the e.g., fields? How should lists return the display value? How to access request parameters (like after one of those security enhancements)? Like a living coding documentation: What is the current reference implementation (and that one covered by unit-tests).

Right now @bpabiszczak or @mariuszkrzaczkowski may know it, because they talked about it simply in the office, but nobody else knows it (or not before studying code changes in the last xx days).

**Comment by bpabiszczak** [comment]
@migoi everything you said is true! We see our mistakes, our flaws, and shortcomings. Even inside the company we keep discussing whether or not what we do is good because many users might stumble upon some problems because of that.

Unfortunately Vtiger itself was poorly written [and at the very beginning of the project we had no idea, we, ourselves, were just starting to figure programming out], and even though 3 years have passed we still keep fixing their low quality code. At the start we decided that the first 2-3 years were the right time to introduce changes quickly, and changes that “break” the system, because once the system gets more popular then many companies would suffer. So right now we’re at the end of that road that end users might consider “inconvenient”. Keep in mind that a ton of projects, like SuiteCRM, SugarCRM, Vtiger, JoForce, EPESI, or coreBOS have awful code. It can’t be changed without affecting companies that use the system, and the more time passes the worse it gets. Looking at it from this perspective, I think we made the right decision as far as these changes are concerned.

Additionally, since Feb 2017, together with another company, we’ve been working on one of the largest deployments in Europe for more than 12k users. Because of this, for a few months, we didn’t have anybody who’d be able to develop the open source version the way we wanted it to be – we are still building our dev teams. A year ago we were a 7 people team, right now there are 18 of us and even more will join in Dec. You have to give us some time to build a “professional” release cycle, we know how to do it, we just need the right people. Starting from 5.0 we want to release hotfixes on an ongoing basis for all fixed errors.

Of course "developer chat hour" is a good idea [and we have a ton of other good ideas] and we will most likely make that happen, but the question is why hasn't the community ever bought the 15h package and spent it on "developer chat hours". And - will just an hour be enough... Because from my experience 1h is usually what one person needs, not the whole group.

**Comment by PercyP** [comment]
Hi @migoi, @bpabiszczak etc.
I have a number of customisations which get effected by the updates but for me it is worth the effort as I know its all going in the right direction. As they say, "you have to crack a few eggs to make an omelette" we are (hopefully) nearly finished 'cracking eggs'. 
I like the bold decisions that have been taken to address the bad code early on - even if it has caused headaches and confusion at times. When things are going smoothly, it can either be because the product is so perfect it no longer has errors/bugs (rarely ever the case) or it is not getting updated or moving forward and has become stagnant!
It looks like Yetiforce is growing and establishing itself, these things take time. 
I like the rapid releases (and all the new features they bring). I also appreciate the work that goes into it. 

> Even inside the company we keep discussing whether or not what we do is good because many users might stumble upon some problems because of that.

When this does happen, the support here has been relatively good. Thankfully everything I have personally come across in terms of update issues has been easily fixed, and I have always taken a backup AND run the update first on a copy of my master version.

**Comment by zebrahosting** [comment]
Yes there has been troubles, we have had to do many new re-installs, got pretty nasty reactions on simple things and then again tested and tested many others and still come back to Yetiforce since it still is our favourite and heading in a solid feature. 

The hotfix idea has been around since v2 but not really started unless something was really broken. We would love to see this implemented and good it is on the list again. 
We also fully understand the hurdles that are there with the growing pain. Thats why we are still here from day one put the extra time in too re-install, fix, report and re-import things.
 
We recommended Yeiforce CRM to our clients but got in trouble when they paid voor support but never got a support person to answer their questions properly. Bizar situation. The whole paid support idea should work but not if the same support department is overloaded. We also paid for support we did not fully use but there is no way to check how much is left. The whole paid support system should be setup a bit better if that is a real plan to make money.

On our wishlist a way to export / import customs fields and/or custom modules . Reason: with the frequent new installs to avoid bringing errors to new releases we have to redo them every time and makes it the most time consuming issue.

Long story short: lets keep working together to make Yetiforce better. Big steps are done, more steps and trouble ahead but it is worth it. Lets try to keep the communication friendly and positive. (In ancient Taoist tradition: with 'soft eyes').

**Comment by PercyP** [comment]
@zebrahosting 
I too like the hotfix concept. It would be easier to track individual issues with updates. I also like the idea of being able to export import custom fields and module. I think we can export/import custom modules already? I have not yet tried this - so not sure if this is the case.

**Comment by KatarzynaUlichnowska** [comment]
The update package has been improved. You can find it [here](https://github.com/YetiForceCompany/UpdatePackages/tree/developer/YetiForce%20CRM%204.x.x/4.2.0_to_4.3.0/zip).

**Comment by migoi** [comment]
@KatarzynaUlichnowska isn't there too many fles below updates/files/vendor/ e.g. ./composer?
Seems they are taken from real installation

**Comment by mariuszkrzaczkowski** [comment]
@migoi I checked the files and they are correct, what is missing from you?

**Comment by mariuszkrzaczkowski** [comment]
@migoi I checked the files and they are correct, what is missing from you?

**Comment by migoi** [comment]
@mariuszkrzaczkowski the Certs, the installed.json. That may depend whats installed on the local installation? Or do i mix up a development and a production fileset

**Comment by kpaulaha** [comment]
If RC1 run earlier RC2 not installing on that. But suppose no need to.
![update](https://user-images.githubusercontent.com/10330264/33484166-ceb4b1ca-d6a9-11e7-980c-8ff8960b4e32.png)

**Comment by PercyP** [comment]
Just run the update - so far so good. All custom modules/fields still working. Menus working great. WIll continue testing

**Comment by paula-w** [comment]
Could you please open a new ticket for this (and any other) issue? It would help us keep track of all the issues. Please send your full logs to github@yetiforce.com, and paste the fragment with the error into the issue.  Thanks a lot!

**Comment by bpabiszczak** [comment]
@zebrahosting I’m not entirely sure what you mean when it comes to support, we register all hours we spend on helping the client [from what I see you used 13h 40min last year]. I don’t remember what your issues were, and which ones we managed to solve and which we didn’t, so I can’t talk about that. If you think any of the issues are still open you can write directly to p.walenczak@yetiforce.com and she will take care of them.

I know for sure that if our cooperation had been closer and more intensive, YetiForce would’ve been a better product today, so remember that it’s not completely up to us what YetiForce is like, the clients and community can influence that as well.

**Comment by paula-w** [comment]
The package has been updated to version RC4.

**Comment by bpabiszczak** [comment]
We upgraded a few of our clients’ systems from some old versions to 4.3 RC and we found a some errors in the old update packages. All packages where the errors existed were fixed and can be found here: https://github.com/YetiForceCompany/UpdatePackages
Next week [before Christmas] we will publish 4.3 and if there are any comments or if you found any errors please report them by Tuesday, because from Wednesday on we are going to fix errors for 4.4 already.

**Comment by mariuszkrzaczkowski** [comment]
Unfortunately, we have not been able to fix all the important errors,  so we are moving the date of publication 4.3 to 2017-12-29.
A new update package 4.2.0_to_4.3.0_RC8.zip is ready https://github.com/YetiForceCompany/UpdatePackages/tree/developer/YetiForce%20CRM%204.x.x/4.2.0_to_4.3.0

**Comment by PercyP** [comment]
Thanks for letting us know. Just about to go test the new update :O)
Have a great holiday

**Comment by zzaz1** [comment]
how update 4.2 to 4.3
site /index.php?module=ModuleManager&parent=Settings&view=ModuleImport&mode=importUserModuleStep1

not work

**Comment by zzaz1** [comment]
Incorrect module version. Module version: 4.2.0 - System version: 4.3.0
